fft/src/KokkosFFT_Cuda_types.hpp

```diff
@@ -15,27 +22,95 @@ static_assert(alignof(cufftComplex) <= alignof(Kokkos::complex<float>));
 static_assert(sizeof(cufftDoubleComplex) == sizeof(Kokkos::complex<double>));
 static_assert(alignof(cufftDoubleComplex) <= alignof(Kokkos::complex<double>));
 
-#ifdef ENABLE_HOST_AND_DEVICE
-#include <fftw3.h>
-#include "KokkosFFT_utils.hpp"
-static_assert(sizeof(fftwf_complex) == sizeof(Kokkos::complex<float>));
-static_assert(alignof(fftwf_complex) <= alignof(Kokkos::complex<float>));
-
-static_assert(sizeof(fftw_complex) == sizeof(Kokkos::complex<double>));
-static_assert(alignof(fftw_complex) <= alignof(Kokkos::complex<double>));
-#endif
-
 namespace KokkosFFT {
 namespace Impl {
 using FFTDirectionType = int;
 
-// Unused
-template <typename ExecutionSpace>
-using FFTInfoType = int;
+/// \brief A class that wraps cufft for RAII
+struct ScopedCufftPlan {
+ private:
+  cufftHandle m_plan;
+
+ public:
+  ScopedCufftPlan(const Kokkos::Cuda &exec_space, int nx, cufftType type,
+                  int batch) {
+    try {
+      cufftResult cufft_rt = cufftPlan1d(&m_plan, nx, type, batch);
+      KOKKOSFFT_THROW_IF(cufft_rt != CUFFT_SUCCESS, "cufftPlan1d failed");
+      set_stream(exec_space);
+    } catch (const std::runtime_error &e) {
+      std::cerr << e.what() << std::endl;
+      cleanup();
+      throw;
+    }
```

Comment:
```suggestion
    cufftResult cufft_rt = cufftPlan1d(&m_plan, nx, type, batch);
    KOKKOSFFT_THROW_IF(cufft_rt != CUFFT_SUCCESS, "cufftPlan1d failed");
    try {
      set_stream(exec_space);
    } catch (const std::runtime_error &e) {
      std::cerr << e.what() << std::endl;
      cleanup();
      throw;
    }
```

We don't want to cleanup if `cufftPlan1d` fails.

This should work indeed but I think this shows that we try to do too many things in the constructor thus making it complicated to read and reason about. I think we should separate the plan creation from the stream installation, like they do in SYCL they have a `commit` function that takes the `sycl::queue` right ?

Reply:
> We don't want to cleanup if cufftPlan1d fails.

Sure

> This should work indeed but I think this shows that we try to do too many things in the constructor thus making it complicated to read and reason about. I think we should separate the plan creation from the stream installation, like they do in SYCL they have a commit function that takes the sycl::queue right ?

The problem is that the lifetime of scoped plan instance is identical to the lifetime of Plan class, since we need to keep it for potential execution of the plan.

https://github.com/kokkos/kokkos-fft/blob/95bd3b317201087e73cc4bf9786158ef22087383/fft/src/KokkosFFT_Plans.hpp#L111-L112

Thus, even if scoped plan instance is constructed, its destructor will not be called if the destructor of Plan class is not called (this is the case if the Plan class construction is disrupted by exception). 
In principle, we have the situation like this
https://godbolt.org/z/veMhKsfjh

If `commit` function throws, we will have the same issue. Thus, I think we need to cleanup by ourselves.

Reply:
If we remove the `try catch`, I got some errors which lead to memory leaks.
https://godbolt.org/z/8KrfKh81a

Reply:
This is actually expected, if there is no try/catch block, there is no unwinding and the program can terminate immediately. It is the responsibility of the user of Kokkos-fft to put the code in a try/catch block.

Reply:
So your point is that

1. w/o try catch
If the `scoped plan` construction fails, the program will be terminated anyway, thus we are good

2. w/ try catch
If the `scoped plan` construction fails, the scoped plan will be destroyed appropriately. 

Then, I am fine to move the logics in the constructor to `commit` method

Reply:
Yes I think that's it. But in my opinion we don't even have to think about whether the constructor is called inside a try/catch block. We have to ensure that:
- if the constructor does not complete, all dynamic allocations are freed without relying on the destructor
- if it does complete, the destructor will free all dynamic allocations.

In a C++ binding over a C object oriented library we can almost have an automatic approach. We can wrap all C opaque objects in a corresponding C++ class that will only call the `*_create` function in the constructor and the `*_destroy` function in the destructor. The rest can, and I think should, be done in extra member functions in order to keep the logic simple in the constructor.

What do you think ?

Reply:
For example gtkmm, a C++ binding to gtk, is almost automatically generated by a set of rules making the link between the C and C++ objects.

Reply:
I mostly agree with you.

> Yes I think that's it. But in my opinion we don't even have to think about whether the constructor is called inside a try/catch block. We have to ensure that:
> 
> * if the constructor does not complete, all dynamic allocations are freed without relying on the destructor

If plan construction fails, we do not need to destroy it because it is not created? 
Then, it can be achieved except for `rocfft`. For `rocfft`, we need a special cleanup method to destroy two different objects (`plan` and `info`).

> * if it does complete, the destructor will free all dynamic allocations.

Sure. It can be easily implemented.

Do you agree that we need a special logic for `rocfft`?